selfservice/flow/registration/handler.go

```diff
@@ -318,29 +318,79 @@ type createBrowserRegistrationFlow struct {
 	if err != nil {
-		h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+		h.d.SelfServiceErrorManager().Forward(ctx, w, r, err)
 		return
 	}
 
-	if sess, err := h.d.SessionManager().FetchFromRequest(r.Context(), r); err == nil {
-		if r.URL.Query().Has("login_challenge") {
-			logoutUrl := urlx.AppendPaths(h.d.Config().SelfPublicURL(r.Context()), logout.RouteSubmitFlow)
-			self := urlx.CopyWithQuery(
-				urlx.AppendPaths(h.d.Config().SelfPublicURL(r.Context()), RouteInitBrowserFlow),
-				r.URL.Query(),
-			).String()
-
-			http.Redirect(
-				w,
-				r,
-				urlx.CopyWithQuery(logoutUrl, url.Values{
-					"token":     {sess.LogoutToken},
-					"return_to": {self},
-				}).String(),
-				http.StatusFound,
-			)
+	var (
+		hydraLoginRequest   *hydraclientgo.OAuth2LoginRequest
+		hydraLoginChallenge sqlxx.NullString
+	)
+	if r.URL.Query().Has("login_challenge") {
+		var err error
+		hydraLoginChallenge, err = hydra.GetLoginChallengeID(h.d.Config(), r)
+		if err != nil {
+			h.d.SelfServiceErrorManager().Forward(ctx, w, r, err)
+			return
+		}
+
+		hydraLoginRequest, err = h.d.Hydra().GetLoginRequest(ctx, hydraLoginChallenge.String())
+		if err != nil {
+			h.d.SelfServiceErrorManager().Forward(ctx, w, r, err)
+			return
+		}
+
+		// on OAuth2 flows, we need to use the RequestURL
+		// as the ReturnTo URL.
+		// This is because a user might want to switch between
+		// different flows, such as login to registration and login to recovery.
+		// After completing a complex flow, such as recovery, we want the user
+		// to be redirected back to the original OAuth2 login flow.
+		if hydraLoginRequest.RequestUrl != "" && h.d.Config().OAuth2ProviderOverrideReturnTo(r.Context()) {
+			q := r.URL.Query()
+			// replace the return_to query parameter
+			q.Set("return_to", hydraLoginRequest.RequestUrl)
+			r.URL.RawQuery = q.Encode()
+		}
+	}
+
+	if sess, err := h.d.SessionManager().FetchFromRequest(ctx, r); err == nil {
+
+		if hydraLoginRequest != nil {
+			// hydra indicates that we cannot skip the login request
+			// so we must perform the login flow.
+			if !hydraLoginRequest.GetSkip() {
+				hydraUrl := hydraLoginRequest.RequestUrl
+				http.Redirect(
+					w,
+					r,
+					hydraUrl,
+					http.StatusFound,
+				)
+				return
+			}
+
+			rt, err := h.d.Hydra().AcceptLoginRequest(r.Context(),
+				hydra.AcceptLoginRequestParams{
+					LoginChallenge:        string(hydraLoginChallenge),
+					IdentityID:            sess.IdentityID.String(),
+					SessionID:             sess.ID.String(),
+					AuthenticationMethods: sess.AMR,
+				})
+			if err != nil {
+				h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+				return
+			}
+			returnTo, err := url.Parse(rt)
+			if err != nil {
+				h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to parse URL: %s", rt)))
+				return
+			}
+			x.AcceptToRedirectOrJSON(w, r, h.d.Writer(), err, returnTo.String())
```

Comment:
I don't think we ever need to return JSON here.

Reply:
We do, in the case of an SPA we return 422 with a browser redirect URL - just like in the Login flow.

Reply:
https://github.com/ory/kratos/blob/master/selfservice/flow/login/handler.go#L479-L496